lib/executor/helpers.go

```diff
@@ -84,26 +87,53 @@ func getIterationRunner(
 ) func(context.Context, lib.ActiveVU) bool {
 	return func(ctx context.Context, vu lib.ActiveVU) bool {
 		err := vu.RunOnce()
+		state := vu.GetState()
 
 		// TODO: track (non-ramp-down) errors from script iterations as a metric,
 		// and have a default threshold that will abort the script when the error
 		// rate exceeds a certain percentage
 
 		select {
 		case <-ctx.Done():
-			// Don't log errors or emit iterations metrics from cancelled iterations
+			// TODO: Use enum?
+			// TODO: How to distinguish interruptions from signal (^C)?
+			tags := map[string]string{"cause": "duration"}
+			state.Samples <- stats.Sample{
+				Time:   time.Now(),
+				Metric: metrics.InterruptedIterations,
+				Tags:   stats.IntoSampleTags(&tags),
+				Value:  1,
+			}
 			executionState.AddInterruptedIterations(1)
 			return false
 		default:
 			if err != nil {
-				if s, ok := err.(fmt.Stringer); ok {
+				// TODO: Use enum?
+				tags := map[string]string{"cause": "error"}
+
+				// TODO: investigate context cancelled errors
+				switch e := err.(type) {
+				case *goja.Exception:
+					if er, ok := e.Value().Export().(lib.IterationInterruptedError); ok {
+						tags["cause"] = er.Cause()
+					}
+					logger.Error(err.Error())
+				case fmt.Stringer:
```

Comment:
This literally was worded as `fmt.Stringer` because I didn't want to add one more place where lib and lib/executor specifically depend on goja. So I think it is better if we don't actually start using it and also you need to fix the fact that now exceptions don't have "source=stacktrace" ;) 

I do think now is a good time to actually add and error type (under lib), probably just an interface "Exception", that is implemented in the `js` package and wraps goja.Exception so that we don't use this directly.

Reply:
Hhmm sure, I'll give that a try, thanks.

Reply:
I managed to remove the goja dependency in d52efc49, but I'm not happy with how it turned out...

I wanted to treat all errors and exceptions consistently, avoid logging in `runFn()` and delegate that to `getIterationRunner()`, but this took an embarassing amount of trial/error and I'm still not sure all errors will be handled correctly. :-/

Some notes:
- I didn't think making `lib.Exception` an interface was needed, as it's generic enough to have a single implementation.
- This needs a lot more tests to ensure errors are rendered properly.
- The `source=stacktrace` logger field doesn't make sense to me. If anything the source should be the same as the "cause" value, as "stacktrace" isn't really a "source".
- Should all errors output the full stack trace? I was trying to keep it consistent with `master`, but maybe errors from `fail()` should only output a short single-line trace with the last stack frame.